File: web/src/solver/validate.ts
/**
 * 解の妥当性検証(pyref/atlasopt/validate.py の移植)。
 *
 * ソルバーの「最適性証明」は渡したモデルに対するものでしかないため、
 * 返ってきた解が元の問題の実行可能解であることを必ず独立に確認する。
 */

import { cost, reachableFrom, type AtlasGraph } from "../data/graph";
import type { SolveResult } from "./result";

/** 問題点のリストを返す(空なら妥当)。 */
export function validate(
  g: AtlasGraph,
  result: SolveResult,
  terminals: Iterable<string>,
  excluded: Iterable<string> = [],
): string[] {
  const problems: string[] = [];
  const sel = new Set(result.nodes);
  const excludedSet = new Set(excluded);
  const wanted = new Set(terminals);
  wanted.delete(g.root);

  if (!sel.has(g.root)) {
    problems.push("root not in solution");
  }

  const hit = [...sel].filter((n) => excludedSet.has(n)).sort();
  if (hit.length > 0) {
    problems.push(`solution contains excluded nodes: ${hit.join(", ")}`);
  }

  const ignored = new Set(result.ignoredTerminals);
  const missing = [...wanted].filter((t) => !ignored.has(t) && !sel.has(t)).sort();
  if (missing.length > 0) {
    problems.push(`terminals missing from solution: ${missing.join(", ")}`);
  }

  // 無視されたterminalは本当に到達不能か(過剰に無視していないか)
  const reachable = reachableFrom(g.adj, g.root, excludedSet);
  const bogus = [...ignored].filter((t) => reachable.has(t) && !excludedSet.has(t)).sort();
  if (bogus.length > 0) {
    problems.push(`terminals ignored despite being reachable: ${bogus.join(", ")}`);
  }

  // 連結性: 選択ノードの誘導部分グラフ上で root から全選択ノードに届くか
  if (sel.size > 0) {
    const banned = new Set([...g.adj.keys()].filter((n) => !sel.has(n)));
    const reach = reachableFrom(g.adj, g.root, banned);
    const same = reach.size === sel.size && [...sel].every((n) => reach.has(n));
    if (!same) {
      problems.push("solution is not connected");
    }
  }

  let expectedPoints = 0;
  for (const n of sel) expectedPoints += cost(g, n);
  if (result.points !== expectedPoints) {
    problems.push(`points mismatch: reported ${result.points}, recomputed ${expectedPoints}`);
  }

  return problems;
}

File: pyref/atlasopt/validate.py
"""解の妥当性検証。

ソルバーの「最適性証明」は渡したモデルに対するものでしかないため、
返ってきた解が元の問題の実行可能解であることを必ず独立に確認する。
"""

from __future__ import annotations

from .graph import AtlasGraph, reachable_from
from .result import SolveResult


def validate(g: AtlasGraph, result: SolveResult, terminals, excluded=()) -> list[str]:
    """問題点のリストを返す(空なら妥当)。"""
    problems: list[str] = []
    sel = set(result.nodes)
    excluded = set(excluded)
    wanted = set(terminals) - {g.root}

    if g.root not in sel:
        problems.append("root not in solution")

    hit = sel & excluded
    if hit:
        problems.append(f"solution contains excluded nodes: {sorted(hit)}")

    ignored = set(result.ignored_terminals)
    missing = wanted - ignored - sel
    if missing:
        problems.append(f"terminals missing from solution: {sorted(missing)}")

    # 無視されたterminalは本当に到達不能か(過剰に無視していないか)
    reachable = reachable_from(g.adj, g.root, banned=frozenset(excluded))
    bogus = {t for t in ignored if t in reachable and t not in excluded}
    if bogus:
        problems.append(f"terminals ignored despite being reachable: {sorted(bogus)}")

    # 連結性: 選択ノードの誘導部分グラフ上で root から全選択ノードに届くか
    banned = frozenset(set(g.adj) - sel)
    if sel and reachable_from(g.adj, g.root, banned=banned) != sel:
        problems.append("solution is not connected")

    expected_points = sum(g.cost(n) for n in sel)
    if result.points != expected_points:
        problems.append(f"points mismatch: reported {result.points}, recomputed {expected_points}")

    return problems
